packages/gatsby/cache-dir/loader.js

```diff
@@ -1,166 +1,152 @@
-              } else {
-                delete jsonPromiseStore[resourceName]
-                reject()
-              }
-            }
-          }
-          req.send(null)
-        })
-        jsonPromiseStore[resourceName] = fetchPromise
-        return fetchPromise
-      }
+      foundPath = trimmedPathname
     }
   }
-
-  // Download the resource
-  hasFetched[resourceName] = true
-  return new Promise(resolve => {
-    const fetchPromise = resourceFunction()
-    let failed = false
-    return fetchPromise
-      .catch(() => {
-        failed = true
-      })
-      .then(component => {
-        fetchHistory.push({
-          resource: resourceName,
-          succeeded: !failed,
-        })
-
-        fetchHistory = fetchHistory.slice(-MAX_HISTORY)
-
-        resolve(component)
-      })
-  })
+  cleanAndFindPathCache[trimmedPathname] = foundPath
+  return foundPath
 }
 
-const prefetchResource = resourceName => {
-  if (resourceName.slice(0, 12) === `component---`) {
-    return Promise.all(
-      createComponentUrls(resourceName).map(url => prefetchHelper(url))
-    )
+const cachedFetch = (resourceName, fetchFn) => {
+  if (resourceName in fetchPromiseStore) {
+    return fetchPromiseStore[resourceName]
   } else {
-    const url = createJsonURL(jsonDataPaths[resourceName])
-    return prefetchHelper(url)
+    const promise = fetchFn(resourceName)
+    fetchPromiseStore[resourceName] = promise
+    return promise.catch(err => {
+      delete fetchPromiseStore[resourceName]
+      return err
+    })
   }
 }
 
-const getResourceModule = resourceName =>
-  fetchResource(resourceName).then(preferDefault)
+const doFetch = url =>
+  new Promise((resolve, reject) => {
+    const req = new XMLHttpRequest()
+    req.open(`GET`, url, true)
+    req.withCredentials = true
+    req.onreadystatechange = () => {
+      if (req.readyState == 4) {
+        resolve(req)
+      }
+    }
+    req.send(null)
+  })
 
-const appearsOnLine = () => {
-  const isOnLine = navigator.onLine
-  if (typeof isOnLine === `boolean`) {
-    return isOnLine
+const handlePageDataResponse = (path, req) => {
+  fetchedPageData[path] = true
+  if (req.status === 200) {
+    const contentType = req.getResponseHeader(`content-type`)
+    if (!contentType || !contentType.startsWith(`application/json`)) {
+      pageNotFoundPaths.add(path)
+      return null
+    } else {
+      const pageData = JSON.parse(req.responseText)
+      pageDatas[path] = pageData
+      return pageData
+    }
+  } else if (req.status === 404) {
+    pageNotFoundPaths.add(path)
+    return null
+  } else {
+    throw new Error(`error fetching page`)
   }
+}
 
-  // If no navigator.onLine support assume onLine if any of last N fetches succeeded
-  const succeededFetch = fetchHistory.find(entry => entry.succeeded)
-  return !!succeededFetch
+const fetchPageData = path => {
+  const url = makePageDataUrl(path)
+  return cachedFetch(url, doFetch).then(req =>
+    handlePageDataResponse(path, req)
```

Comment:
It's subtle, but [during prefetch](https://github.com/gatsbyjs/gatsby/pull/13409/files/b5798bc4d53f75c3ec708747bbc6c2375b7f58fc#diff-06ce5193de7c2c01dd947a5001c147c3R235) we need to make a full XHR which will actually be served from the browser's cache since it was just prefetched. So `fetchPageData` is designed to be the low level function, and [loadPageData](https://github.com/gatsbyjs/gatsby/pull/13409/files/b5798bc4d53f75c3ec708747bbc6c2375b7f58fc#diff-06ce5193de7c2c01dd947a5001c147c3R256) is the higher level function that caches the response.